### ISSUE TITLE
NSGAIIStoppingByTimeRunner Experiment in JMetal?

### ISSUE THREAD

**Issue by momah** [jMetal/jMetal#373]
I am performing **experiments** with JMetal and I have to say that metal's experiments are very useful, but my interest is to use **StoppingByTime criteria** instead of maxEvaluations.

We imagine that we want to perform the NSGAIIStudy-like experiments, but instead of using NSGAII we want to use NSGAIIStoppingByTime and that each instance lasts 1 minute(`int thresholdComputingTimeInMilliseconds = 60000;`)
- NSGAIIStudy -> https://github.com/jMetal/jMetal/blob/master/jmetal-exec/src/main/java/org/uma/jmetal/experiment/NSGAIIStudy.java
- NSGAIIStoppingByTime -> https://github.com/jMetal/jMetal/blob/master/jmetal-algorithm/src/main/java/org/uma/jmetal/algorithm/multiobjective/nsgaii/NSGAIIStoppingByTime.java

It is easy to change the calls of the NSGAII algorithm to NSGAIIStoppingByTime in the experiments template, But..

### **TROUBLE**

**After the execution, the duration of the entire experiment** of all instances and INDEPENDENT_RUNS  only lasts (1 minute = 60000 ms) as long as I want each of the INDEPENDENT_RUNS to last.

**It seems that the system does not recognize that each instance has its independent thresholdComputingTimeInMilliseconds variable.**

In other words, if I want to experiment with 3 algorithms, 25 INDEPENDENT_RUNS and each run lasts 1 minute, the experiment should last **3X25X1 = 75 minutes** (or a little less for parallelism, I guess), however, the experiment only **lasts 1 minute in total** (the thresholdComputingTimeInMilliseconds value)

**CODE EXAMPLE**

![image](https://user-images.githubusercontent.com/2429022/67876889-89c67f00-fb30-11e9-9039-fb91ebfe2524.png)

Thanks in advance.
mmahrach

**Comment by ajnebro** [comment]
Hi @momah.
  Very interesting issue. You are using a feature, stopping algorithms by time, in the context of an experiment. I have never tried that, so it is likely that something is not working properly.

  Let me think about it.

**Comment by matthieu-vergne** [comment]
Since you know it should run in total around 75 minutes, you may set the limit to 75 minutes. Of course it is not a proper solution, but it seems to me to be the closest and simplest workaround.

The issue with time computation is that you cannot guarantee the determinism. We do have some randomization processes, but we can use the same seeds to ensure we obtain the same results. With time computation, we cannot, since the computation effort is distributed among threads and other processes on the machine based on the decision of the processor, which may vary between runs. In particular, time limits are always approximative since the processors itself cannot guarantee to stop exactly at the requested nanosecond. Moreover, the machine may be overloaded at some point, but the impacted runs will nevertheless stop after the given time.

Although it does not completely discard the use of time limits, it does support the fact that we cannot have reliable experiments with them. In other words, even if we do solve the issue you highlighted here, at best I expect it to be just a bit more precise than the workaround I suggest at the beginning of this post. So the effort to fix it might be unworthy.

**Comment by momah** [comment]
Thank you very much for your answers.

Regarding the option to set the limit in the total expected time (in the previous example 75 min), it does not execute all the executions (INDEPENDENT_RUNS) simultaneously, I have noticed that only the first 8 executions run for the entire time, the rest for a less time, and some don't even run.

**I think it is due to the order of executions established by the method (`myProblem.setNumberOfCores (8);`)**

![image](https://user-images.githubusercontent.com/2429022/68022722-0d04e380-fc9d-11e9-8616-a5cf974457fd.png)

Sincerely I don't know what this method does (if they are threads / logical cores,..). But i have solved this by changing the parameter to 75 `myProblem.setNumberOfCores (75)`, where 75 is the total of the INDEPENDENT_RUNS (3X25X1).

After changing setNumberOfCores from 8 to 75 all INDEPENDENT_RUNS are executed simultaneously, but I do not know if setNumberOfCores are threads of execution or what really happens in the machine.

**Conclusions:** 
**I don't know whether to trust the results in this case, and in the end I decided to do the experiments using (maxEvaluations corresponding to one minute for each run INDEPENDENT_RUN).**

**The inconvenience in this case i think, is that the time it takes for an algorithm to perform a number of evaluations can be very different from another. Any suggestions in this regard?**

Anyway thank you very much.

**Comment by matthieu-vergne** [comment]
> The inconvenience in this case i think, is that the time it takes for an algorithm to perform a number of evaluations can be very different from another. Any suggestions in this regard?

The most reliable suggestion I can give is "don't use time limits". Even if we fix all of that, as mentioned before, we cannot give any guarantee that one run of 1 min will perform a number of evaluations equal, not even close, to another run of 1 min. Time limits are a good limit for approximative evaluations, in order to have an idea of the number of evaluations it can perform per unit of time. Once you have this idea, you can establish the number of evaluations you are ready to wait for, but don't expect to produce any reliable experiment with time limits. Never trust results based on time limits, they are never reproducible.

This issue for time limits is a generic one, not just for our case. Whatever software you run on whatever hardware, time limits never guarantee strict reproducibility because of random variables. If having a comparable number of evaluations is important for you, then your best choice is to run for a given amount of evaluations.

I don't minimize this issue, which clearly shows problems in our code. But if your concern is the comparability, which is a good thing, then time limits are recognised as bad practices.

**Comment by momah** [comment]
I will consider your recommendation. Thanks again to the whole team for their fast support and great tool.
